expo/expo-cli
# [RFC] esbuild support

# Why

esbuild is a bundler written with native code, react-native is almost exclusively used with Metro bundler. This PR demonstrates how esbuild could be implemented at scale for cross platform React apps. Based on some work I did in mid-2020 -- [tweet](https://twitter.com/Baconbrix/status/1361439672086392832?s=20).

> This is a side project and not official Expo work. Expo is a small team without the bandwidth to pursue every project.

# How

Initially I built the system using a fork of React Native that had flow stripped out. Since then, esbuild has added a plugin system which can manually strip flow (suboptimal as every project must do this).

Most of the plugin system was assembled by @dalcib in his project [exbuild](https://github.com/dalcib/exbuild), symbolication is based on [re.pack](https://github.com/callstack/repack/blob/3c1e0597557d09ab64fab0a29a86d3c487a23ef0/packages/repack/src/server/Symbolicator.ts). 

The purpose of this PR is to show users how modular the bundler system is in Expo CLI, and how a new bundler can be added.

## Planned

- [x] Native support
- [x] Expo Go / Dev Client support
- [x] Asset support
- [x] Websocket controls (press R and M to interact with the client)
- [x] Symbolication (error stack traces)
  - [x] Open to code
  - [x] Code preview 
- [x] Hot module reloading
- [ ] Flipper 
  - [x] React Dev Utils inspector
- [ ] Hermes (maybe done)
- [x] Bare support -- `EXPO_BUNDLER=esbuild expo start --dev-client`
- [ ] [Fast Refresh](https://github.com/evanw/esbuild/issues/645)
- [ ] Production builds
- [ ] Web support
- [ ] Hidden file output
- [ ] Monorepo support (should be able to start expo/apps/native-component-list)
- [ ] Integrated Expo logging

# Test Plan

Run the following in an Expo project:

```
EXPO_BUNDLER=esbuild expo start
```

And open the project in the Expo Go app on iOS. Use `EXPO_PLATFORM=android` to bundle for Android (should improve this long term).

- Make some changes and save the project to see HMR.
- Press `r`, `m`, `shift+m` in the terminal to interact with the client.

## Stack Traces

Throw an error in the project to test symbolication, similar traces should be collapsed in esbuild. There is a known bug right now where the LogBox doesn't format the symbols unless you tap to the next error/warning.
